lightning-data-persister/src/lib.rs

```diff
@@ -0,0 +1,239 @@
+extern crate lightning;
+extern crate bitcoin;
+extern crate libc;
+
+use bitcoin::hash_types::{BlockHash, Txid};
+use bitcoin::hashes::hex::{ToHex, FromHex};
+use lightning::chain::channelmonitor::{ChannelMonitor, ChannelMonitorUpdate, ChannelMonitorUpdateErr};
+use lightning::chain::keysinterface::ChannelKeys;
+use lightning::chain::transaction::OutPoint;
+use lightning::ln::data_persister::ChannelDataPersister;
+use lightning::util::ser::{Writeable, Readable};
+use std::collections::HashMap;
+use std::fs;
+use std::io::{Error, ErrorKind, Cursor};
+use std::marker::PhantomData;
+use std::path::Path;
+
+#[cfg(not(target_os = "windows"))]
+use std::os::unix::io::AsRawFd;
+
+/// FilesystemPersister can persist channel data on disk on Linux machines, where
+/// each channel's data is stored in a file named after its funding outpoint.
+///
+/// Warning: this module does the best it can with calls to persist data, but it
+/// can only guarantee that the data is passed to the drive. It is up to the
+/// drive manufacturers to do the actual persistence properly, which they often
+/// don't (especially on consumer-grade hardware). Therefore, it is up to the
+/// user to validate their entire storage stack, to ensure the writes are
+/// persistent.
+/// Corollary: especially when dealing with larger amounts of money, it is best
+/// practice to have multiple channel data backups and not rely only on the
+/// FilesystemPersister.
+pub struct FilesystemPersister<ChanSigner: ChannelKeys + Readable + Writeable> {
+	path_to_channel_data: String,
+	phantom: PhantomData<ChanSigner>, // TODO: is there a way around this?
```

Comment:
You can drop the parameter entirely. impl blocks can have their own generic requirements that are separate from the type they're impl'ing for.

Reply:
Hm that gives me `unconstrained type parameter`. Here's what else I tried:
1) adding the traits to the `Keys` type parameter for the `DataPersister` trait, and that gave me `only auto traits can be used as additional traits in a trait object`.
2) adding the type parameter to each individual method of the `DataPersister` trait, which gave me `the trait ln::data_persister::ChannelDataPersister cannot be made into an object ...because method update_channel_data has generic type parameters`. 

That said, I'm probably missing something small lol. Lmk what you think.

Reply:
Ah! Right, so I know way too much about this cause I spent about a full day fighting with Rust in the C bindings on the same issue - if you define a trait with an associated type (instead of templating the trait on the type explicitly with "<>"), then the associated type in an impl block must have a single concrete type for the associated type (vs when doing explicit templates, the compiler will effectively generate a new impl block for each different template parameter). Thus, we have to make the FilesystemPersister templated on ChanSigner (forcing the compiler to generate a new impl block for each "different" FilesystemPersister<X> instead of it trying, and failing, to generate one for FilesysytemPersister). Because you end up with nonsense unreadable error messages blowing up in your face, I dont think we should use associated types (what a shame, they read so much nicer, and the drawback where the compiler ends up generating 20 copies of each impl block doesn't apply to us - users aren't going to have ten different ChannelKeys types flying around).

Indded, changing to a non-associated-type fixes the issue - https://0bin.net/paste/9-oYDqGg#QSbVoeuaMKTtKfgQ4tu5o-r75dOdh6fAhG8UGMTnICH

Reply:
Long conversation on IRC about this, but the useful takeaway is "associated types are for outputs, template types are for inputs" - in this case the ChanSigner is a type that is used in a parameter and thus should be a template. If it were used to a return value that'd be different.

Reply:
Ahh, I didn't know traits could have templated types. Thanks, that's much nicer.